tool/autoqa/wikidata/csqa-converter.js

```diff
@@ -0,0 +1,398 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// This file is part of Genie
+//
+// Copyright 2019-2020 The Board of Trustees of the Leland Stanford Junior University
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Naoki Yamamura <yamamura@cs.stanford.edu>
+"use strict";
+
+
+const fs = require('fs');
+const os = require('os');
+const _ = require('lodash');
+const util = require('util');
+const path = require('path');
+const assert = require('assert');
+
+const ThingTalk = require('thingtalk');
+const Tp  = require('thingpedia');
+const Ast = ThingTalk.Ast;
+const I18N = require('../../../lib/i18n');
+const tokenizer = I18N.get('en-US' ).getTokenizer();
+
+import { serializePrediction } from '../../../lib/utils/thingtalk';
+import {
+    getItemLabel,
+    argnameFromLabel,
+} from './utils';
+import { makeDummyEntities } from "../../../lib/utils/entity-utils";
+
+const QUESTION_TYPES = new Set(
+    ['Simple Question (Direct)',
+    'Verification (Boolean) (All)',
+    'Logical Reasoning (All)',
+    'Quantitative Reasoning (All)',
+    'Quantitative Reasoning (Count) (All)',
+    'Comparative Reasoning (All)',
+    'Comparative Reasoning (Count) (All)']);
+
+class CsqaConverter {
+    constructor(options) {
+        this._domains = options.domains;
+        this._canonicals = options.canonicals;
+        this._input_dir = options.inputDir;
+        this._output_dir = options.outputDir;
+        this._dataset = options.dataset;
+        this._instances = options.instances;
+        this._propertyLabels = options.propertyLabels;
+        this._entityLabels = options.entityLabels;
+        this._pathes;
+    }
+
+    async _readSync(func, dir) {
+        return util.promisify(func)(dir, { encoding: 'utf8' });
+    }
+
+    async _filterDomainQAPairs(canonical) {
+        const basePath = path.join(this._input_dir, this._dataset)
+        let cnt = 0;
+        const qaPairs = [];
+        for (const dir of (await this. _readSync(fs.readdir, basePath))) {
+            for (const file of (await this. _readSync(fs.readdir, path.join(basePath, dir)))) {
+                const dialog = JSON.parse((await this. _readSync(fs.readFile, path.join(basePath, dir, file))));
+                let userTurn;
+                for (const turn in dialog) {
+                    const speaker = dialog[turn].speaker;
```

Comment:
there are quite some nested loop/conditions, probably consider create a separate function such as `filterTurnByDomain` to make this function easier to read